### ISSUE TITLE
CakePHP 3, associated data NOT being saved nor updated.

### ISSUE THREAD

**Issue by efloriani** [cakephp/cakephp#8865]
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 3
* Platform and Target: UBUNTO MYSQL Your version of PHP is 5.5.9 or higher (detected 5.5.9-1ubuntu4.16).

### What you did
Hello,

Im trying to edit a associated table and it does not happen. I have Countries that has many States, that has many Cities, that has many Companis. 

Here are my tables:

```sql
--Countries
CREATE TABLE IF NOT EXISTS `countries` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `name` varchar(45) NOT NULL,
  `created` datetime NOT NULL,
  `modified` datetime NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `name_UNIQUE` (`name`)
) ENGINE=InnoDB  DEFAULT CHARSET=utf8 AUTO_INCREMENT=3 ;

--
--  `countries` table DATA:
--

INSERT INTO `countries` (`id`, `name`, `created`, `modified`) VALUES
(1, 'Brasil', '2016-05-18 18:21:44', '2016-05-18 18:21:44'),
(2, 'Estados Unidos', '2016-05-18 18:21:50', '2016-05-18 18:21:50');

-- States Table:
CREATE TABLE IF NOT EXISTS `states` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `country_id` int(11) NOT NULL,
  `name` varchar(45) NOT NULL,
  `abbreviation` varchar(2) NOT NULL,
  `created` varchar(45) NOT NULL,
  `modified` varchar(45) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `manual_UNIQUE` (`country_id`,`name`),
  UNIQUE KEY `manual_UNIQUE2` (`country_id`,`abbreviation`),
  KEY `manual_idx` (`id`,`country_id`)
) ENGINE=InnoDB  DEFAULT CHARSET=utf8 AUTO_INCREMENT=7 ;

--
-- `states` table DATA:
--

INSERT INTO `states` (`id`, `country_id`, `name`, `abbreviation`, `created`, `modified`) VALUES
(1, 1, 'Santa Catarina', 'SC', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM'),
(2, 1, 'Paraná', 'PR', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM'),
(3, 1, 'Rio Grande do Sul', 'RS', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM'),
(4, 2, 'New York', 'NY', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM'),
(5, 2, 'Washington', 'WS', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM'),
(6, 2, 'Texas', 'TX', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM');

--Cities Table
CREATE TABLE IF NOT EXISTS `cities` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `state_id` int(11) NOT NULL,
  `state_country_id` int(11) NOT NULL,
  `name` varchar(45) NOT NULL,
  `created` varchar(45) NOT NULL,
  `modified` varchar(45) NOT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `manual_UNIQUE` (`state_id`,`state_country_id`,`name`),
  KEY `fk_cities_states1_idx` (`state_id`,`state_country_id`),
  KEY `manual_idx` (`id`,`state_id`,`state_country_id`)
) ENGINE=InnoDB  DEFAULT CHARSET=utf8 AUTO_INCREMENT=14 ;

--
-- `cities` table DATA:
--

INSERT INTO `cities` (`id`, `state_id`, `state_country_id`, `name`, `created`, `modified`) VALUES
(1, 2, 1, 'Curitiba', '5/18/16, 6:22 PM', '5/19/16, 2:49 PM'),
(2, 2, 1, 'São José dos Pinhais', '5/18/16, 6:22 PM', '5/18/16, 6:22 PM'),
(3, 4, 2, 'New York', '5/18/16, 6:23 PM', '5/19/16, 2:52 PM'),
(4, 5, 2, 'Washington', '5/18/16, 6:23 PM', '5/19/16, 2:53 PM'),
(5, 6, 2, 'Cidade do México', '5/18/16, 6:23 PM', '5/18/16, 6:23 PM'),
(6, 3, 1, 'Porto Alegre', '5/18/16, 6:23 PM', '5/18/16, 6:23 PM'),
(7, 3, 1, 'Canoas', '5/18/16, 6:23 PM', '5/19/16, 2:51 PM'),
(8, 1, 1, 'Blumenau', '5/18/16, 7:44 PM', '5/19/16, 2:48 PM'),
(9, 1, 1, 'Gaspar', '5/18/16, 7:44 PM', '5/18/16, 7:44 PM'),
(10, 1, 1, 'Jaraguá do Sul', '5/19/16, 12:46 PM', '5/19/16, 12:46 PM'),
(11, 1, 1, 'São Bento do Sul', '5/19/16, 12:56 PM', '5/19/16, 12:56 PM'),
(12, 4, 2, 'Another City', '5/19/16, 1:03 PM', '5/19/16, 2:49 PM'),
(13, 6, 2, 'Texas City', '5/19/16, 1:14 PM', '5/19/16, 2:54 PM');

--Companies Table:
CREATE TABLE IF NOT EXISTS `companies` (
  `id` int(11) NOT NULL AUTO_INCREMENT,
  `city_id` int(11) DEFAULT NULL,
  `city_state_id` int(11) DEFAULT NULL,
  `city_state_country_id` int(11) DEFAULT NULL,
  `name` varchar(255) NOT NULL,
  `created` datetime NOT NULL,
  `modified` datetime NOT NULL,
  `code` varchar(45) NOT NULL,
  `address` varchar(45) DEFAULT NULL,
  `zip_code` decimal(11,0) DEFAULT NULL,
  `phone` decimal(11,0) DEFAULT NULL,
  `website` varchar(45) DEFAULT NULL,
  `companiescol` varchar(45) DEFAULT NULL,
  PRIMARY KEY (`id`),
  UNIQUE KEY `code_UNIQUE` (`code`),
  UNIQUE KEY `name_UNIQUE` (`name`),
  KEY `fk_companies_cities1_idx` (`city_id`,`city_state_id`,`city_state_country_id`)
) ENGINE=InnoDB  DEFAULT CHARSET=utf8 AUTO_INCREMENT=5 ;

--
-- `companies` table DATA:
--

INSERT INTO `companies` (`id`, `city_id`, `city_state_id`, `city_state_country_id`, `name`, `created`, `modified`, `code`, `address`, `zip_code`, `phone`, `website`, `companiescol`) VALUES
(1, 2, 2, 1, 'Empresa teste2', '2016-05-19 13:29:44', '2016-05-19 13:29:44', '454645', '', NULL, NULL, '', ''),
(2, 5, 6, 2, 'New Company', '2016-05-19 13:34:25', '2016-05-19 13:34:25', '456456456465', '', NULL, NULL, '', ''),
(3, 6, 3, 1, 'TREI FILIAL', '2016-05-19 13:53:25', '2016-05-19 13:53:25', '041110', 'Rua Norberto Achterberg, 170', 89032241, 4791889393, 'www.trei.com.br', 'descritivos'),
(4, 6, 3, 1, 'Segunda Empresa Porto', '2016-05-20 14:46:26', '2016-05-20 14:46:26', '4564564564564', '', NULL, NULL, '', '');
```

-- Companies Table Model:

```php
parent::initialize($config);

        $this->table('companies');
        $this->displayField('name');
        $this->primaryKey('id');

        $this->addBehavior('Timestamp');

        /*$this->belongsTo('Cities', [
            'foreignKey' => 'city_id'
        ]);*/
		
        $this->belongsTo('States', [ //Modificado originalmente de Cities para States, vinculando assim e podendo adicionar o objeto States com nome completo no controller
            'foreignKey' => 'city_state_id'
        ]);
        $this->belongsTo('Countries', [ //Modificado originalmente de Cities para Countries, vinculando assim e pondendo adicionar o objeto Countries com nome completo no controller
            'foreignKey' => 'city_state_country_id'
        ]);
		
		$this->belongsTo('Cities', [
			'bindingKey' => ['id', 'state_id', 'state_country_id'],
			'foreignKey' => ['city_id', 'city_state_id', 'city_state_country_id'],
		]);
```

--Cities Table model:
```php
parent::initialize($config);

        $this->table('cities');
        $this->displayField('name');
        $this->primaryKey('id');

        $this->addBehavior('Timestamp');

        $this->belongsTo('States', [
            'foreignKey' => 'state_id',
            'joinType' => 'INNER'
        ]);
        $this->belongsTo('Countries', [ //Modificado originalmente do backup, Aqui estava States e vinculava errado os estados, e agora vincula com a tabela Countries
            'foreignKey' => 'state_country_id',
            'joinType' => 'INNER'
        ]);
		/*$this->hasMany('Companies', [
				'bindingKey' => ['id', 'state_id', 'state_country_id'],
				'foreignKey' => ['city_id', 'city_state_id', 'city_state_country_id'],
				]);*/
	
		
        $this->hasMany('Companies', [
			//'bindingKey' => 'id',
            'foreignKey' => 'city_id'
        ]);
```

--Cities Controller (EDIT):

```php
public function edit($id = null)
    {
        $city = $this->Cities->get($id, [
            'contain' => ['Companies']
        ]);
		
        if ($this->request->is(['patch', 'post', 'put'])) {
			
			//debug($city) or die('aqui');
			//debug($this->request->data()) or die('aqui');
            $city = $this->Cities->patchEntity($city, $this->request->data(), [
				'associated' => ['Companies']
			]);
			
			//debug($city) or die('aqui');
			$conn = \Cake\Datasource\ConnectionManager::get('default');
			$conn->logQueries(true);
			/*$this->Cities->save($city, [
					'associated' => ['Companies']
				]);*/
			try { //Com try e catch possível tratar os erros.
				if ($this->Cities->save($city, [
					'atomic'=>false,
					'validate' => false,
					'associated' => ['Companies']
				])) {
					$this->Flash->success(__('The city has been saved.'));
					return $this->redirect(['action' => 'index']);
				} else {
					$this->Flash->error(__('The city could not be saved. Please, try again.'));
				}
			}
			catch (\PDOException $e) {
				$error = 'The item you are trying to delete is associated with other records';
				// The exact error message is $e->getMessage();
				$this->Flash->error($e->getMessage());
				print $e->getMessage() or die();
			}
				$conn->logQueries(false);
        }
        //$states = $this->Cities->States->find('list', ['limit' => 200]);
		
		/*
		 * Abaixo estamos solicitando todos os estados do País que a cidade está atualmente configurada no banco de dados e listando todos para seleção.
		 */
		$states = $this->Cities->States->find('list', array(
				'conditions' => array('States.country_id' => $city->state_country_id),
				'recursive' => -1));
		/*
		 * Abaixo estamos selecionando todos os países para que possa ser novamente alterado uma cidade de país, caso tenha sido cadastrada equivocadamente.
		 */
		$countries = $this->Cities->States->Countries->find('list', ['limit' => 200]);
        $this->set(compact('city', 'states','countries'));
        $this->set('_serialize', ['city']);
    }
```

--Cities edit.ctp:

```php
<?= $this->Form->create($city) ?>
    <fieldset>
        <legend><?= __('Edit City') ?></legend>
        <?php
			echo $this->Form->input('state_country_id', ['options' => $countries]);
            echo $this->Form->hidden('companies.0.id');
			echo $this->Form->input('companies.0.city_state_id', ['options' => $states]);
			echo $this->Form->hidden('companies.1.id');
			echo $this->Form->input('companies.1.city_state_id', ['options' => $states]);
			echo $this->Form->input('state_id', ['options' => $states]);
			
            echo $this->Form->input('name');
			

        ?>
    </fieldset>
    <?= $this->Form->button(__('Submit')) ?>
    <?= $this->Form->end() ?>
```

### Expected Behavior
In this example, Im trying to change the state_id from 3 to 2 ON the city id=6, because this city has two companies.

For now I just put in form view edit.ctp manually input for 2 companies, but I could have a lot of companies in that city on my table. So if I change this City State, I would like that the system change not only the city state_id (Cities table) but also change all companies city_state_id (Companies table)

### Actual Behavior
Actually when I try to save, the system just try to do a update on cities table on save() with associated table:
2016-05-20 16:58:40 Debug: duration=71 rows=0 UPDATE cities SET state_id = 2 , modified = '5/20/16, 4:58 PM' WHERE id = 6

SQLSTATE[23000]: Integrity constraint violation: 1451 Cannot delete or update a parent row: a foreign key constraint fails (`trei_preprod`.`companies`, CONSTRAINT `fk_companies_cities1` FOREIGN KEY (`city_id`, `city_state_id`, `city_state_country_id`) REFERENCES `cities` (`id`, `state_id`, `state_country_id`) ON DE)

**Comment by efloriani** [comment]
-- Request data from form looks like: 
[
	'state_country_id' => '1',
	'companies' => [
		(int) 0 => [
			'id' => '3',
			'city_state_id' => '2'
		],
		(int) 1 => [
			'id' => '4',
			'city_state_id' => '2'
		]
	],
	'state_id' => '2',
	'name' => 'Porto Alegre'
]

-- $city After patching with associated: 
object(App\Model\Entity\City) {

	'id' => (int) 6,
	'state_id' => (int) 2,
	'state_country_id' => (int) 1,
	'name' => 'Porto Alegre',
	'created' => '5/18/16, 6:23 PM',
	'modified' => '5/18/16, 6:23 PM',
	'companies' => [
		(int) 0 => [
			'id' => '3',
			'city_state_id' => '2'
		],
		(int) 1 => [
			'id' => '4',
			'city_state_id' => '2'
		]
	],
	'[new]' => false,
	'[accessible]' => [
		'*' => true,
		'id' => true
	],
	'[dirty]' => [
		'companies' => true,
		'state_id' => true
	],
	'[original]' => [
		'companies' => [
			(int) 0 => object(App\Model\Entity\Company) {

				'id' => (int) 3,
				'city_id' => (int) 6,
				'city_state_id' => (int) 3,
				'city_state_country_id' => (int) 1,
				'name' => 'TREI FILIAL',
				'created' => object(Cake\I18n\FrozenTime) {

					'time' => '2016-05-19T13:53:25+00:00',
					'timezone' => 'UTC',
					'fixedNowTime' => false
				
				},
				'modified' => object(Cake\I18n\FrozenTime) {

					'time' => '2016-05-19T13:53:25+00:00',
					'timezone' => 'UTC',
					'fixedNowTime' => false
				
				},
				'code' => '041110',
				'address' => 'Rua Norberto Achterberg, 170',
				'zip_code' => (float) 89032241,
				'phone' => (float) 4791889393,
				'website' => 'www.trei.com.br',
				'companiescol' => 'descritivos',
				'[new]' => false,
				'[accessible]' => [
					'*' => true,
					'id' => true
				],
				'[dirty]' => [],
				'[original]' => [],
				'[virtual]' => [],
				'[errors]' => [],
				'[invalid]' => [],
				'[repository]' => 'Companies'
			
			},
			(int) 1 => object(App\Model\Entity\Company) {

				'id' => (int) 4,
				'city_id' => (int) 6,
				'city_state_id' => (int) 3,
				'city_state_country_id' => (int) 1,
				'name' => 'Segunda Empresa Porto',
				'created' => object(Cake\I18n\FrozenTime) {

					'time' => '2016-05-20T14:46:26+00:00',
					'timezone' => 'UTC',
					'fixedNowTime' => false
				
				},
				'modified' => object(Cake\I18n\FrozenTime) {

					'time' => '2016-05-20T14:46:26+00:00',
					'timezone' => 'UTC',
					'fixedNowTime' => false
				
				},
				'code' => '4564564564564',
				'address' => '',
				'zip_code' => null,
				'phone' => null,
				'website' => '',
				'companiescol' => '',
				'[new]' => false,
				'[accessible]' => [
					'*' => true,
					'id' => true
				],
				'[dirty]' => [],
				'[original]' => [],
				'[virtual]' => [],
				'[errors]' => [],
				'[invalid]' => [],
				'[repository]' => 'Companies'
			
			}
		],
		'state_id' => (int) 3
	],
	'[virtual]' => [],
	'[errors]' => [],
	'[invalid]' => [],
	'[repository]' => 'Cities'

}

**Comment by lorenzo** [comment]
What do you actually expect it happens? you are telling cake to update state_id, and it did... but the database is telling you that it is not possible based on your foreign key restrictions. What would be the correct solution to you?

**Comment by thinkingmedia** [comment]
This should be closed. Please post questions for support on stackoverflow.

https://stackoverflow.com/questions/tagged/cakephp

**Comment by markstory** [comment]
If your operations violate your constraints, there isn't much CakePHP can do to solve that.

**Comment by efloriani** [comment]
Hello Lorenzo,

What Im expecting: When I save this object $city above, I want cake to update the state_id on cities table but also at Companies table. At Companies table I have a city_state_id that I could patch sucefuly after a long time trying. As you can see abot the $city object now has 2 companies that should have their city_state_id also updated.

**Comment by efloriani** [comment]
Hello Markstory.

Companies table has a city_id, city_state_id and city_state_country_id. 

When I have a new company, I select country, state, and city for that company. When I edit the city state, like I did something wrong and want to change the city state on CITIES table, I get this constraints error. I was expecting that cake would save at once all $city object that has all interactions that should be done with the tables. When I change the state id for this city in CITIES table, I was expecting that the cakephp change the citi_state_id at Companies table, so I wouldn´t have problems any more. It was dificult but the object $city seams fine, and brings all companies for this particular city that I changed the state_id.
